### ISSUE TITLE
'window' is undefined in SSR

### ISSUE THREAD

**Issue by wilfredmifsud** [lentschi/ngx-ellipsis#27]
Hi

While testing your plugin, I am encountering an issue where a `window is not defined` error is showing while rendering the website in SSR.

From what I can see you have no checks if the `window` exists in the current context.

It currently shows the text "undefined" instead of the actual text.

![image](https://user-images.githubusercontent.com/739777/66922273-059bd400-f027-11e9-93bb-c44e531121d5.png)

As an alternative, I tried using `ellipsis-resize-detection="manual"` which is suppose to use no resizing listeners, but it still shows the text as `undefined` instead of the actual text.

I would really like it to work somehow in SSR and non-SSR version of the website.

Any help would be great

**Comment by lentschi** [comment]
@wilfredmifsud You're right, thanks for bringing this up!

It's fixed in version 2.0.8, which I just published.

What I did:

1. `resize-detection-strategy="element-resize-detector"` (previous default setting) won't work at startup, since it requires the `window` object to be present. So I changed the logic to make the value default to `window` in that case (`@HostListener('window:resize')` still does work.) If you really need to use `element-resize-detector` though, I found that wrapping the ellipsis element in an `*ngIf` does the trick (It will then be initialized in the client instead of on the server)
2. I use `HTMLElement`'s `innerText` to retrieve the original ellipsis content. With SSR there is no such property as it is [apparently non-standard](https://stackoverflow.com/questions/35213147/difference-between-textcontent-vs-innertext#answer-35213639). So I changed it to still use `innerText` if available (to stay compatible with older non-SSR versions), but fallback on the standardized `textContent`.

**Comment by wilfredmifsud** [comment]
Thats great.. I pulled the latest version and it does not break in SSR.

Will be testing it further on 

What I would suggest as an improvement on top of this, is to apply `overflow-y: hidden` when in SSR mode, thus clipping the text (I know it is not the same as clipping dynamically and applying '...' at the end) so that the difference between SSR and browser is minimal.

**Comment by lentschi** [comment]
> What I would suggest as an improvement on top of this, is to apply overflow-y: hidden

@wilfredmifsud Hm... I'm not sure... As described in the README, `ngx-ellipsis` only takes care of the truncating bit. How you ascertain that the element ends up with the size you want it to have, is up to you. (In the simplest case it's something like `overflow: hidden; width: 100px; height: 100px;` but you could for example also make it a flex element with it's overflow being cut off by a neighbouring absolutely positioned element.)
Personally I dislike libs that auto-attach element styles, because that is difficult to override if required due to some special need - so I would rather not add that to the lib, sorry.
If you need that throughout your app, you could just set a single global CSS rule, to achieve it:

```css
*[ellipsis] {
  overflow: hidden;
}
```

Also I don't see why this is a point about SSR - it's the same as in the browser - But maybe I'm not getting something there...? :confused: 

Finally I would like to point out one other thing (You may already know this, but someone else stumbling across this ticket might wonder about it):

One of the things Angular Universal was made for is to accelerate initial page display. (So you can see some content before the page becomes interactive.) So though version `ngx-ellipsis` 2.0.8 did fix some SSR initialization errors, it doesn't comply with the acceleration aspect at all: To calculate the letters that will fit inside the ellipsis element, the lib uses `HTMLElement` properties such as `clientWidth` and `scrollWidth`, which are unavailable on the server side (AFAIK Angular Universal cannot now how exactly text will be rendered on the OS and browser the client uses)
**So in short:** The ellipsis is not being being precalculated on the server and it **will slow down the client**. If anyone has an idea on how to prevent that, I'm all ears :wink:

**Comment by lentschi** [comment]
@wilfredmifsud Btw. I just released yet another version (`2.0.9`), because my last fix was a bit ugly. (In the new version the detection default is back to `element-resize-detector` - it just won't try to execute the code leading up to the resize detection before we're in the browser - I was wrong to just disable it)

**Comment by kamiour** [comment]
Hi, @lentschi 
The issue seems to be reproduced once again on version ^3.1.3
I get this when compiling in SSR mode:
const ResizeObserver = window.ResizeObserver || _juggle_resize_observer__WEBPACK_IMPORTED_MODULE_1__["ResizeObserver"];
                       ^
ReferenceError: window is not defined

![image](https://user-images.githubusercontent.com/52451228/119819242-d5056700-bef8-11eb-9159-aaac220b6a2c.png)

**Comment by lentschi** [comment]
@kamiour Yeah, sorry - fixed in `ngx-ellipsis@3.1.4`

(I missed this when switching the resize observer to  `@juggle/resize-observer`, which was required for #58.)

**Comment by kamiour** [comment]
@lentschi thanks a lot!
